### PR TITLE
Added settings for DEBUGINFOD cache location and timeout

### PR DESCRIPTION
**Pull request by kevinfrei**
I've been working on more/better configuration for improving DEBUGINFOD support. This is the first (and easiest) slice of the work.

I've added `timeout` and `cache-path` settings that can override the DEBUGINFOD library defaults (and environment variables.) I also renamed the `plugin.symbol-locator.debuginfod.server_urls` setting to `server-urls` to be more consistent with the rest of LLDB's settings (the underscore switch is switched to a hyphen)

I've got a few tests that validate the cache-path setting (as a side-effect), but they've exposed a few bugs that I'll be putting up a separate PR for (which will include the tests).

### CODE REVIEW

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:58]: Is it always safe to return a StringRef here? It looks fine at first sight, but this code also doesn't look particularly hot, so playing it safe and returning a `std::string` avoids having to reason about the string's lifetime.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:150]: LLVM's code style requires a period at the end of comments. Same across the patch.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:155]: This comment is stating the obvious. The interesting part is that this is stored in the plugin options, which may be worth mentioning here.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:169]: `url_path`

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:170]: `cache_key`

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:58]: String storage lives 40 lines lower (line 97) because, yeah, this bit me. The Debuginfod library takes a StringRef, and I found this was the most consistent way to deal with it. If you'd prefer me to switch this and cons up a StringRef before the API, I'm okay with that (just let me know in here)

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:169]: What's the style guidance for locals? The locals above this are filled with PascalCaseOrErr everywhere. I'm happy to change the entire function, but right now it's really inconsistent (and kinda ugly). I'd like to make it consistent, either snake or pascal, based on your guidance, here. Should I just snake them all (cache_directory_path_or_err?) or Pascal them all (BuildId), or is there some rule I'm missing that makes the mixture sensible?

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:170]: ditto from above.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:169]: Yeah, LLDB diverges from LLVM's coding style in this: https://lldb.llvm.org/resources/contributing.html#coding-style. it should use snake_case everywhere.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:155]: Modified to describe the 'interesting part'.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:177]: LLDB_LOGV (LLDBLog::Symbols context) seems like the right choice, given how noisy it will be.

File: llvm/include/llvm/Debuginfod/Debuginfod.h
Comment (mysterymath) [inline llvm/include/llvm/Debuginfod/Debuginfod.h:49]:
```suggestion
/// Returns the cache key for a given debuginfod URL path.
```

File: llvm/include/llvm/Debuginfod/Debuginfod.h
Comment (mysterymath) [inline llvm/include/llvm/Debuginfod/Debuginfod.h:64]:
```suggestion
/// Get the full URL path for a source request of a given BuildID and file path.
```

File: llvm/include/llvm/Debuginfod/Debuginfod.h
Comment (mysterymath) [inline llvm/include/llvm/Debuginfod/Debuginfod.h:73]:
```suggestion
/// Get the full URL path for an executable request of a given BuildID.
```

File: llvm/include/llvm/Debuginfod/Debuginfod.h
Comment (mysterymath) [inline llvm/include/llvm/Debuginfod/Debuginfod.h:80]:
```suggestion
/// Get the full URL path for a debug binary request of a given BuildID.
```

**Review by mysterymath — COMMENTED** [review]
LGTM for debuginfod

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:60]: Actually, I do think it's wrong after all. On line 69 you return a `StringRef` to a temporary on the stack (created on line 66). This must be a std::string.

**Review by JDevlieghere — CHANGES_REQUESTED** [review]

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:60]: D'oh. You're right. I'll hoist that StringRef up and switch the function to return a string. Good catch!

**Comment by kevinfrei** [comment]
> LGTM for debuginfod

I've considered moving that entire thing into a namespace. Would you be amenable to that (at a later date...)?

Unimportant: The OMF2097 github pic is pretty excellent, but clearly you're almost as old as me...

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td:12]: I think this is rather weird: why not make 90 the default? If I specify zero I'd expect no timeout.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:178]: I'd drop the `[Debuginfod]` as it's not something other plugins do and you can see where the log message came from which will make it obvious that that's coming from the debuginfod plugin. If you really want to include it, I'd do something like "Debuginfod failed".

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td:12]: Both of these settings exist to let lldb users override values that are configurable through environment variables used in the Debuginfod library. 0 means "use the DEBUGINFOD_TIMEOUT value" which then defaults to a timeout of 90 seconds.

What does "no timeout" mean? Like, just wait forever? I'm not sure that's an option in curl... I could switch the default value to -1, just to be more clearly a nonsensical value. @clayborg also seemed confused by this option without more explanation, so I think it needs a better description. How about this:

Timeout (in seconds) for requests made to a DEBUGINFOD server. A value of zero [could be -1 instead] indicates that the system default should be used, which is separately configurable through the DEBUGINFOD_TIMEOUT environment variable.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:178]: > I'd drop the `[Debuginfod]` as it's not something other plugins do and you can see where the log message came from which will make it obvious that that's coming from the debuginfod plugin. If you really want to include it, I'd do something like "Debuginfod failed".

I can't see that it's coming from the Debuginfod plugin: 
```
(lldb) log enable lldb symbol --verbose
(lldb) target create /bin/tar
Debuginfod failed to download symbol artifact buildid/b4e36170d228ca4ae032f06b16f69adb48d07c50/debuginfo with error curl_easy_perform() failed: Timeout was reached
```

so I'm inclined to just switch it to...what you can see I switched it to. If you want me to create a new LLDBLog category, I do believe *that* would make it more obvious. I just reused LLDBLog::Symbols.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td
Comment (JDevlieghere) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td:12]: Gotcha, that makes sense and that's not how I understood it when I read the description. Not sure if this is better but maybe something along the lines of: 

> Timeout (in seconds) for requests made to a DEBUGINFOD server. A value of zero means we use the debuginfod default timeout: DEBUGINFOD_TIMEOUT if the environment variable is set and 90 seconds otherwise.

Comment (kevinfrei) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td:12]: I make no claims to my ability to write readable prose. Your description makes at least much as sense as mine, so I'll just swipe it directly.

File: lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp
Comment (JDevlieghere) [reply lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp]: Yeah you need to pass `-F` to `log enable` to include the file and function names. Anyway this works for me!

**Review by JDevlieghere — APPROVED** [review]
Thanks! This LGTM.

**Review by clayborg — COMMENTED** [review]
Looks good to me on the LLDB side, but you should probably get the ok from the maintainers of the debug infod library for those changes.

**Comment by kevinfrei** [comment]
> Looks good to me on the LLDB side, but you should probably get the ok from the maintainers of the debug infod library for those changes.

I think that was @mysterymath

**Review by mysterymath — APPROVED** [review]

**Pull request merged**
